File: content/en/04-llms/llms/_exercises/24_shinychat-1/_agent.R
list_skills <- function(skills_dir) {
  files <- fs::dir_ls(skills_dir, recurse = TRUE, glob = "**/SKILL.md")
  skills <- lapply(files, \(path) frontmatter::read_front_matter(path)$data)

  paste(
    vapply(
      skills,
      \(skill) paste0("- ", skill$name, ": ", skill$description),
      character(1)
    ),
    collapse = "\n"
  )
}

prep_blockbuster_agent <- function(client, project_dir, skills_dir) {
  path_for <- function(path) file.path(project_dir, path)

  read_file <- function(path) {
    brio::read_file(path_for(path))
  }

  write_file <- function(path, content) {
    full_path <- path_for(path)
    fs::dir_create(fs::path_dir(full_path))
    brio::write_file(content, full_path)
    paste0("Wrote ", path, ".")
  }

  list_files <- function() {
    files <- fs::dir_ls(project_dir, recurse = TRUE, type = "file")
    paste(fs::path_rel(files, start = project_dir), collapse = "\n")
  }

  edit_file <- function(path, old, new) {
    full_path <- path_for(path)
    content <- brio::read_file(full_path)
    matches <- gregexpr(old, content, fixed = TRUE)[[1]]

    if (identical(matches, -1L)) {
      stop("Could not find the text to replace in ", path, ".")
    }

    if (length(matches) != 1L) {
      stop(
        "Expected one exact match in ",
        path,
        ", but found ",
        length(matches),
        "."
      )
    }

    brio::write_file(sub(old, new, content, fixed = TRUE), full_path)
    paste0("Edited ", path, ".")
  }

  read_skill <- function(skill) {
    brio::read_file(file.path(skills_dir, skill, "SKILL.md"))
  }

  system_prompt <- paste(
    "You are a coding agent for the Last Blockbuster in Bend, Oregon.",
    "",
    "Read a skill with the read_skill tool before you do the job it describes.",
    "For a new renewal letter, list letters/drafts first and write the next version as",
    "letters/drafts/<member-name>-vN.md.",
    "",
    "## Available skills",
    "",
    list_skills(skills_dir),
    sep = "\n"
  )

  client$set_system_prompt(system_prompt)

  client$register_tool(ellmer::tool(
    read_file,
    description = "Read the full contents of a file in the Blockbuster workspace.",
    arguments = list(
      path = ellmer::type_string("Path relative to the Blockbuster workspace.")
    )
  ))
  client$register_tool(ellmer::tool(
    write_file,
    description = "Write a file in the Blockbuster workspace, overwriting it if it exists.",
    arguments = list(
      path = ellmer::type_string("Path relative to the Blockbuster workspace."),
      content = ellmer::type_string("The full contents of the file to write.")
    )
  ))
  client$register_tool(ellmer::tool(
    list_files,
    description = "List files in the Blockbuster workspace, including draft letters."
  ))
  client$register_tool(ellmer::tool(
    edit_file,
    description = "Replace one exact span of text in a Blockbuster workspace file.",
    arguments = list(
      path = ellmer::type_string("Path relative to the Blockbuster workspace."),
      old = ellmer::type_string("The exact existing text to replace."),
      new = ellmer::type_string("The replacement text.")
    )
  ))
  client$register_tool(ellmer::tool(
    read_skill,
    description = "Read the full instructions for a listed skill before doing its job.",
    arguments = list(
      skill = ellmer::type_string("Name of the skill to read.")
    )
  ))

  invisible(client)
}

File: content/en/04-llms/llms/_solutions/25_shinychat-2/_agent.py
from pathlib import Path

import frontmatter
from chatlas import ChatPosit


def list_skills(skills_dir: Path) -> str:
    skill_files = sorted(skills_dir.rglob("SKILL.md"))
    skills = [frontmatter.load(path).metadata for path in skill_files]
    return "\n".join(
        f"- {skill['name']}: {skill['description']}" for skill in skills
    )


def prep_blockbuster_agent(
    client: ChatPosit,
    project_dir: Path,
    skills_dir: Path,
) -> None:
    def read_file(path: str) -> str:
        """Read the full contents of a file in the Blockbuster workspace."""
        return (project_dir / path).read_text()

    def write_file(path: str, content: str) -> str:
        """Write a file in the Blockbuster workspace, overwriting it if it exists."""
        full_path = project_dir / path
        full_path.parent.mkdir(parents=True, exist_ok=True)
        full_path.write_text(content)
        return f"Wrote {path}."

    def list_files() -> str:
        """List files in the Blockbuster workspace, including draft letters."""
        return "\n".join(
            str(path.relative_to(project_dir))
            for path in sorted(project_dir.rglob("*"))
            if path.is_file()
        )

    def edit_file(path: str, old: str, new: str) -> str:
        """Replace one exact span of text in a Blockbuster workspace file."""
        full_path = project_dir / path
        content = full_path.read_text()
        matches = content.count(old)

        if matches == 0:
            raise ValueError(f"Could not find the text to replace in {path}.")
        if matches != 1:
            raise ValueError(f"Expected one exact match in {path}, but found {matches}.")

        full_path.write_text(content.replace(old, new))
        return f"Edited {path}."

    def read_skill(skill: str) -> str:
        """Read the full instructions for a listed skill before doing its job."""
        return (skills_dir / skill / "SKILL.md").read_text()

    system_prompt = f"""
You are a coding agent for the Last Blockbuster in Bend, Oregon.

Read a skill with the read_skill tool before you do the job it describes.
For a new renewal letter, list letters/drafts first and write the next version as
letters/drafts/<member-name>-vN.md.

## Available skills

{list_skills(skills_dir)}
"""

    client.system_prompt = system_prompt
    for tool in (read_file, write_file, list_files, edit_file, read_skill):
        client.register_tool(tool)
